src/Microsoft.AspNet.Mvc.Core/Internal/FilterCache.cs

```diff
@@ -0,0 +1,211 @@
+        private readonly IFilterProvider[] _filterProviders;
+
+        private InnerCache _currentCache;
+
+        public FilterCache(
+            IActionDescriptorCollectionProvider collectionProvider,
+            IEnumerable<IFilterProvider> filterProviders)
+        {
+            _collectionProvider = collectionProvider;
+            _filterProviders = filterProviders.OrderBy(item => item.Order).ToArray();
+        }
+
+        private InnerCache CurrentCache
+        {
+            get
+            {
+                var current = Volatile.Read(ref _currentCache);
+                var actionDescriptors = _collectionProvider.ActionDescriptors;
+
+                if (current == null || current.Version != actionDescriptors.Version)
+                {
+                    current = new InnerCache(actionDescriptors.Version);
+                    Volatile.Write(ref _currentCache, current);
+                }
+
+                return current;
+            }
+        }
+
+        public IFilterMetadata[] GetFilters(ActionContext actionContext)
+        {
+            var cache = CurrentCache;
+            var actionDescriptor = actionContext.ActionDescriptor;
+
+            CacheEntry entry;
+            if (cache.Entries.TryGetValue(actionDescriptor, out entry))
+            {
+                return GetFiltersFromEntry(entry, actionContext);
+            }
+
+            var items = new List<FilterItem>(actionDescriptor.FilterDescriptors.Count);
+            for (var i = 0; i < actionDescriptor.FilterDescriptors.Count; i++)
+            {
+                items.Add(new FilterItem(actionDescriptor.FilterDescriptors[i]));
+            }
+
+            entry = new CacheEntry(items);
+            ExecuteProviders(actionContext, items);
+
+            var filters = ExtractFilters(items);
+
+            var allFiltersCached = true;
+            for (var i = 0; i < items.Count; i++)
+            {
+                var item = items[0];
+                if (!item.ShouldCache)
+                {
+                    item.Filter = null;
+                    allFiltersCached = false;
+                }
+            }
+
+            if (allFiltersCached)
+            {
+                entry = new CacheEntry(filters);
+            }
+
+            cache.Entries.TryAdd(actionDescriptor, entry);
+            return filters;
+        }
+
+        private IFilterMetadata[] GetFiltersFromEntry(CacheEntry entry, ActionContext actionContext)
+        {
+            Debug.Assert(entry.Filters != null || entry.Items != null);
+
+            if (entry.Filters != null)
+            {
+                return entry.Filters;
+            }
+
+            var items = new List<FilterItem>(entry.Items);
+            for (var i = 0; i < entry.Items.Length; i++)
+            {
+                var item = entry.Items[i];
+                if (item.ShouldCache)
+                {
+                    items.Add(item);
```

Comment:
This seems weird - isn't this adding to the `List` twice (since it's already been added as part of the ctor?)